### ISSUE TITLE
誤差関数の選択について

### ISSUE THREAD

**Issue by kazucmpt** [matsuken92/Qiita_Contents#2]
Qiita_Contents/chainer-vae/chainer-vae.ipynbのIn [47]

rec_loss += F.bernoulli_nll(x, self.decode(z, sigmoid=False)) / (k * batchsize)

ですが、これは再構成誤差なので、入力と出力がどの程度等しいかを表す誤差関数だと認識しています。最も素朴なのはここで最小二乗誤差を使うだと思われます。

そこで、

rec_loss += F.mean_squared_error(x, self.decode(z)) / k

とするのが自然ではないでしょうか。

**Comment by matsuken92** [comment]
@kazucmpt さん
"Auto-Encoding Variational Bayes" Diederik P Kingma, Max Welling  https://arxiv.org/abs/1312.6114
によると、下記のようにどちらの誤差を使う場合も解説されています。今回のMNISTのケースは各ピクセルが0, 1の二値を取ると考えるとlog lossでも利用することが可能です。Gaussianにすると両端からはみ出る恐れもあるため、今回はlog lossにしています。（ただし、ground truthが0,1ではなく[0, 1]のグレースケールなのにこれでいいのかという疑問は私も前から持っていますが、その答えは見つけられていません）

C MLP’s as probabilistic encoders and decoders
　C.1 Bernoulli MLP as decoder
　C.2 Gaussian MLP as encoder or decode

**Comment by kazucmpt** [comment]
理解しました。ありがとうございます。
